File: Hero.py
import random

# lớp Hero: chứa thông tin về Hero
class Hero:
    def __init__(self, cấp, máu, thủ, mẫn, công1, công2, elo = 2000, thắng = 0, hòa = 0, thua = 0):
        self.level = cấp        # mỗi cấp được thêm 1 điểm sp phân bổ vào các thuộc tính
        self.hp = máu           # lượng máu, bị trừ đi mỗi khi bị đánh
        self.defence = thủ      # trừ lực đánh của đối phương
        self.agility = mẫn      # khả năng nhân lực đánh (so chỉ số với đối phương)
        self.attack1 = công1    # lực đánh tối thiểu
        self.attack2 = công2    # lực đánh tối đa
        self.elo = elo
        self.win = thắng
        self.draw = hòa
        self.lose = thua

    def __str__(self):
        total = self.win + self.draw + self.lose
        if total > 0:
            return f'(cấp {self.level}) HP: {self.hp}, DEF: {self.defence:0.1f}, AGI: {self.agility:0.1f}, ATK: {self.attack1:0.1f}-{self.attack2:0.1f}, ELO: {self.elo:0.2f}, Rate: {total} [{self.win * 100 / total:0.1f}%/{self.draw * 100 / total:0.1f}%/{self.lose * 100 / total:0.1f}%]'
        else:
            return f'(cấp {self.level}) HP: {self.hp}, DEF: {self.defence:0.1f}, AGI: {self.agility:0.1f}, ATK: {self.attack1:0.1f}-{self.attack2:0.1f}, ELO: {self.elo:0.2f}, Rate: {total}'

# trả về hệ số elo sau khi trận đấu A vs B có kết quả score (thắng/hòa/thua ~ 1/0.5/0)
def gain(A, B, score, K = 10):
    Qa, Qb = 10 ** (A / 400), 10 ** (B / 400)
    E = K * (score - Qa / (Qa + Qb))
    return A + E, B - E

# cho A đấu với B, trả về kết quả (thắng/hòa/thua ~ 1/0.5/0)
# tối đa 100 lượt đánh (turn)
def vs(A, B, turn = 100):
    hp1, hp2 = A.hp, B.hp
    for i in range(turn):
        hp2 -= max(random.uniform(A.attack1, A.attack2) - B.defence, 0) * A.agility / B.agility
        if hp2 < 0: return 1
        hp1 -= max(random.uniform(B.attack1, B.attack2) - A.defence, 0) * B.agility / A.agility
        if hp1 < 0: return 0
    return 0.5

# tốc độ đánh là 1/a và 1/b, tức là b/ab và a/ab
# A đánh trước, bắt đầu vào các thời điểm 0, b, 2b, 3b,...
# B đánh sau, bắt đầu vào các thời điểm 1, 1+a, 1+2a, 1+3a,...
# kết thúc sau khi cả 2 đủ số turn
# chú ý xử lý trường hợp cả 2 cùng ra chiêu
def vs2(A, B, turn = 100):
    hpA, hpB = A.hp, B.hp
    a, b = 0, 50
    turnA, turnB = 0, 0
    while turnA < turn or turnB < turn:
        if a == b:
            hpB -= max(random.uniform(A.attack1, A.attack2) - B.defence, 0)
            a += B.agility
            turnA += 1
            hpA -= max(random.uniform(B.attack1, B.attack2) - A.defence, 0)
            b += A.agility
            turnB += 1
        elif a < b:
            hpB -= max(random.uniform(A.attack1, A.attack2) - B.defence, 0)
            a += B.agility
            turnA += 1
        else:
            hpA -= max(random.uniform(B.attack1, B.attack2) - A.defence, 0)
            b += A.agility
            turnB += 1
        if hpA < 0 and hpB < 0: return 0.5
        if hpB < 0: return 1
        if hpA < 0: return 0
    else: return 0.5

# cho A và B đấu N trận, cập nhật các hệ số
def attack(A, B, N):
    for i in range(N):
        score = vs2(A, B)
        A.elo, B.elo = gain(A.elo, B.elo, score)
        if score == 1:
            A.win += 1
            B.lose += 1
        elif score == 0:
            A.lose += 1
            B.win += 1
        else:
            A.draw += 1
            B.draw += 1

# nạp dữ liệu từ file
def loadfile():
    pass

# ghi dữ liệu ra file
def savefile():
    pass

# thêm hero mới vào bộ dữ liệu
def add_hero(p1, p2, p3, p4):
    level = p1 + p2 + p3 + p4
    hp = p1 * 321
    de = p2 * 3.1
    ag = p3 * 4.1
    at = p4 * 2.4
    # hp gốc 5000, mỗi level thêm 100
    # at tối đa mỗi level thêm 200 (trung bình tăng 100)
    heroes.append(Hero(level, hp + 5000 + level * 100, de, ag + 100, at, at + 100 + 2 * level))

# thêm mọi phương án hero có tổng điểm thuộc tính là sp (tương đương với level = sp)
def add_hero_level(sp):
    for p1 in range(sp + 1):
        for p2 in range(sp - p1 + 1):
            for p3 in range(sp - p1 - p2 + 1):
                p4 = sp - p1 - p2 - p3
                add_hero(p1, p2, p3, p4)

# chọn ngẫu nhiên count cặp hero cho đấu nhau
def train(count = 1_000_000):
    n = 0
    while n < count:
        A = random.choice(heroes)
        B = random.choice(heroes)
        if (A == B) or (400 <= abs(A.elo - B.elo)): continue
        attack(A, B, 1)
        n = n + 1

# lần lượt cho mọi hero đấu với nhau
def train2(limit = 400):
    for a in range(len(heroes)):
        for b in range(a + 1, len(heroes)):
            if limit > abs(heroes[a].elo - heroes[b].elo):
                attack(heroes[a], heroes[b], 1)

h = []
heroes = []
for i in range(300,301):
    heroes = []
    add_hero_level(i)
    h.extend(random.sample(heroes, min(500, len(heroes))))
    #h.extend(heroes)
heroes = h
#add_hero(0, 0, 0, 0)
heroes = random.sample(heroes, min(200, len(heroes)))
random.shuffle(heroes)
print(f"Tất cả {len(heroes)} heroes ==========================")
for i in range(100):
    train2()
    print(f"Vòng {i}: {max(heroes, key = lambda h: h.elo)}")
print("Xếp theo ELO ===============")
for h in sorted(heroes, key = lambda h: h.elo): print(h)
print("Xếp theo Ti lệ thắng ===============")
for h in sorted(heroes, key = lambda h: (h.win / (h.win + h.draw + h.lose), h.draw / (h.win + h.draw + h.lose))): print(h)
print("Xếp theo Tỉ lệ thua ===============")
for h in sorted(heroes, key = lambda h: (h.lose / (h.win + h.draw + h.lose), h.draw / (h.win + h.draw + h.lose))): print(h)

File: Hero.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <random>
#include <iomanip>

using namespace std;

class Hero {
public:
    int level;
    double hp, defence, agility, attack1, attack2, elo;
    int win, draw, lose;

    Hero(int cấp, double máu, double thủ, double mẫn, double công1, double công2, double elo = 2000, int thắng = 0, int hòa = 0, int thua = 0)
        : level(cấp), hp(máu), defence(thủ), agility(mẫn), attack1(công1), attack2(công2), elo(elo), win(thắng), draw(hòa), lose(thua) {}

    string toString() const {
        int total = win + draw + lose;
        ostringstream oss;
        if (total > 0) {
            oss << "(cấp " << level << ") HP: " << hp << ", DEF: " << fixed << setprecision(1) << defence
                << ", AGI: " << agility << ", ATK: " << attack1 << "-" << attack2
                << ", ELO: " << fixed << setprecision(2) << elo
                << ", Rate: " << total << " [" << fixed << setprecision(1)
                << (win * 100.0 / total) << "%/" << (draw * 100.0 / total) << "%/" << (lose * 100.0 / total) << "%]";
        } else {
            oss << "(cấp " << level << ") HP: " << hp << ", DEF: " << fixed << setprecision(1) << defence
                << ", AGI: " << agility << ", ATK: " << attack1 << "-" << attack2
                << ", ELO: " << fixed << setprecision(2) << elo
                << ", Rate: " << total;
        }
        return oss.str();
    }
};

double gain(double A, double B, double score, double K = 1) {
    double Qa = pow(10, A / 400), Qb = pow(10, B / 400);
    double E = K * (score - Qa / (Qa + Qb));
    return A + E;
}

double vs(const Hero& A, const Hero& B, int turn = 100) {
    double hp1 = A.hp, hp2 = B.hp;
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<> disA(A.attack1, A.attack2);
    uniform_real_distribution<> disB(B.attack1, B.attack2);

    for (int i = 0; i < turn; ++i) {
        hp2 -= max(disA(gen) - B.defence, 0.0) * A.agility / B.agility;
        if (hp2 < 0) return 1;
        hp1 -= max(disB(gen) - A.defence, 0.0) * B.agility / A.agility;
        if (hp1 < 0) return 0;
    }
    return 0.5;
}

// Lưu ý:
// - turn = 100: ưu thế nhất là nâng agi + atk
// - turn = 200: ưu thế nhất là nâng def + atk
// - turn = 300: ưu thế nhất là nâng def + agi + atk
// - turn = 400: ưu thế nhất là nâng def
double vs2(Hero& A, Hero& B, int turn = 100) {
    double hpA = A.hp, hpB = B.hp;
    double a = 0, b = 50;
    int turnA = 0, turnB = 0;
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<> disA(A.attack1, A.attack2);
    uniform_real_distribution<> disB(B.attack1, B.attack2);

    while (turnA < turn || turnB < turn) {
        if (a == b) {
            hpB -= max(disA(gen) - B.defence, 0.0);
            a += B.agility;
            turnA++;
            hpA -= max(disB(gen) - A.defence, 0.0);
            b += A.agility;
            turnB++;
        } else if (a < b) {
            hpB -= max(disA(gen) - B.defence, 0.0);
            a += B.agility;
            turnA++;
        } else {
            hpA -= max(disB(gen) - A.defence, 0.0);
            b += A.agility;
            turnB++;
        }
        if (hpA < 0 && hpB < 0) return 0.5;
        if (hpB < 0) return 1;
        if (hpA < 0) return 0;
    }
    return 0.5;
}

void attack(Hero& A, Hero& B, int N) {
    for (int i = 0; i < N; ++i) {
        double score = vs2(A, B);
        double newEloA = gain(A.elo, B.elo, score);
        double newEloB = gain(B.elo, A.elo, 1 - score);
        A.elo = newEloA;
        B.elo = newEloB;

        if (score == 1) {
            A.win++;
            B.lose++;
        } else if (score == 0) {
            A.lose++;
            B.win++;
        } else {
            A.draw++;
            B.draw++;
        }
    }
}

vector<Hero> heroes;

void add_hero(int p1, int p2, int p3, int p4) {
    int level = p1 + p2 + p3 + p4;
    double hp = p1 * 321;
    double de = p2 * 3.1;
    double ag = p3 * 4.1;
    double at = p4 * 2.4;
    heroes.emplace_back(level, hp + 5000 + level * 100, de, ag + 100, at, at + 100 + 2 * level);
}

void add_hero_level(int sp) {
    for (int p1 = 0; p1 <= sp; ++p1) {
        for (int p2 = 0; p2 <= sp - p1; ++p2) {
            for (int p3 = 0; p3 <= sp - p1 - p2; ++p3) {
                int p4 = sp - p1 - p2 - p3;
                add_hero(p1, p2, p3, p4);
            }
        }
    }
}

void train(int count = 1000000) {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, heroes.size() - 1);

    int n = 0;
    while (n < count) {
        Hero& A = heroes[dis(gen)];
        Hero& B = heroes[dis(gen)];
        if (&A == &B || abs(A.elo - B.elo) >= 400) continue;
        attack(A, B, 1);
        n++;
    }
}

void train2(int limit = 400) {
    for (size_t a = 0; a < heroes.size(); ++a) {
        for (size_t b = a + 1; b < heroes.size(); ++b) {
            if (limit > abs(heroes[a].elo - heroes[b].elo)) {
                attack(heroes[a], heroes[b], 1);
            }
        }
    }
}

void train3(int round = 10) {
    vector<vector<int>> r(round, vector<int>(heroes.size(), -1));

    cout << "Thực hiện " << round << " vòng đấu ==========================" << endl;
    // đấu round vòng
    for (int i = 0; i < round; ++i) {
        // tạo danh sách theo elo giảm dần
        vector<int> index(heroes.size());
        iota(index.begin(), index.end(), 0);
        sort(index.begin(), index.end(), [](int a, int b) {
            return heroes[a].elo > heroes[b].elo;
        });
        cout << "Vòng " << i+1 << ": " << heroes[index[0]].toString() << endl;

        // đấu các hero trong danh sách
        for (int k = 0; k < heroes.size(); ++k)
            if (r[i][index[k]] == -1) {
                int p = k;
                while (true) {
                    p++;
                    bool found = false;
                    for (int t = 0; t < i; ++t)
                        if (r[t][index[k]] == p) {
                            found = true;
                            break;
                        }
                    if (!found) break;
                }
                if (p >= heroes.size()) break;
                r[i][index[k]] = p;
                r[i][index[p]] = k;
                attack(heroes[index[k]], heroes[index[p]], 10);
            }
    }
}

int main() {
    /*
    vector<Hero> h;
    for (int i = 1; i <= 40; ++i) {
        heroes.clear();
        add_hero_level(i);
        random_device rd;
        mt19937 gen(rd());
        shuffle(heroes.begin(), heroes.end(), gen);
        h.insert(h.end(), heroes.begin(), heroes.begin() + min(10, (int)heroes.size()));
    }
    heroes = h;
    random_device rd;
    mt19937 gen(rd());
    shuffle(heroes.begin(), heroes.end(), gen);
    */

    add_hero(0, 0, 0, 0);
    for (int lv = 1; lv <= 45; ++lv)
        add_hero_level(lv);
    /*
    for (int lv = 1; lv <= 120; ++lv) {
        add_hero(lv, 0, 0, 0);
        add_hero(0, lv, 0, 0);
        add_hero(0, 0, lv, 0);
        add_hero(0, 0, 0, lv);
    }
    for (int lv = 1; lv <= 60; ++lv) {
        add_hero(lv, lv, 0, 0);
        add_hero(lv, 0, lv, 0);
        add_hero(lv, 0, 0, lv);
        add_hero(0, lv, lv, 0);
        add_hero(0, lv, 0, lv);
        add_hero(0, 0, lv, lv);
    }
    for (int lv = 1; lv <= 40; ++lv) {
        add_hero(lv, lv, lv, 0);
        add_hero(lv, lv, 0, lv);
        add_hero(lv, 0, lv, lv);
        add_hero(0, lv, lv, lv);
    }        
    for (int lv = 1; lv <= 30; ++lv) {
        add_hero(lv, lv, lv, lv);
    }
    for (int lv = 1; lv <= 20; ++lv) {
        add_hero(lv+lv+lv, lv, lv, lv);
        add_hero(lv+lv, lv+lv, lv, lv);
        add_hero(lv+lv, lv, lv+lv, lv);
        add_hero(lv+lv, lv, lv, lv+lv);
        add_hero(lv, lv+lv+lv, lv, lv);
        add_hero(lv, lv+lv, lv+lv, lv);
        add_hero(lv, lv+lv, lv, lv+lv);
        add_hero(lv, lv, lv+lv+lv, lv);
        add_hero(lv, lv, lv+lv, lv+lv);
        add_hero(lv, lv, lv, lv+lv+lv);
    }
    */

    random_device rd;
    mt19937 gen(rd());
    shuffle(heroes.begin(), heroes.end(), gen);

    cout << "Tất cả " << heroes.size() << " heroes ==========================" << endl;
    for (int i = 1; i <= 10; ++i) {
        train3(100);
        auto maxHero = *max_element(heroes.begin(), heroes.end(), [](const Hero& a, const Hero& b) {
            return a.elo < b.elo;
        });
        cout << "Vòng " << i << ": " << maxHero.toString() << endl;
    }

    cout << "Xếp theo ELO ===============" << endl;
    sort(heroes.begin(), heroes.end(), [](const Hero& a, const Hero& b) {
        return a.elo < b.elo;
    });
    for (const auto& h : heroes) {
        cout << h.toString() << endl;
    }
    
    cout << "Xếp theo Tỉ lệ thắng ===============" << endl;
    sort(heroes.begin(), heroes.end(), [](const Hero& a, const Hero& b) {
        double rateA = a.win / (double)(a.win + a.draw + a.lose);
        double rateB = b.win / (double)(b.win + b.draw + b.lose);
        return rateA < rateB;
    });
    for (const auto& h : heroes) {
        cout << h.toString() << endl;
    }

    cout << "Xếp theo Tỉ lệ thua ===============" << endl;
    sort(heroes.begin(), heroes.end(), [](const Hero& a, const Hero& b) {
        double rateA = a.lose / (double)(a.win + a.draw + a.lose);
        double rateB = b.lose / (double)(b.win + b.draw + b.lose);
        return rateA < rateB;
    });
    for (const auto& h : heroes) {
        cout << h.toString() << endl;
    }


    return 0;
}